File: packages/sdk-go/regressa_test.go
package regressa

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"sync/atomic"
	"testing"
	"time"
)

type captured struct {
	Traces []map[string]interface{} `json:"traces"`
	SDK    map[string]string        `json:"sdk"`
}

func server(t *testing.T, status int, got *[]captured, keys *[]string) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		var c captured
		_ = json.Unmarshal(body, &c)
		*got = append(*got, c)
		*keys = append(*keys, r.Header.Get(KeyHeader))
		w.WriteHeader(status)
		_, _ = w.Write([]byte(`{"accepted":1,"rejected":0,"errors":[]}`))
	}))
}

func TestTraceBuffersAndFlushesWithHeader(t *testing.T) {
	var got []captured
	var keys []string
	srv := server(t, 202, &got, &keys)
	defer srv.Close()
	c := New(Options{APIKey: "rgsa_test_abc", BaseURL: srv.URL, FlushAt: 10, FlushInterval: time.Hour})
	out := "hello"
	c.Trace(Trace{Model: "gpt-4o-mini", Provider: OpenAI, InputMessages: []Message{{Role: "user", Content: "hi"}}, OutputText: &out})
	if err := c.Flush(context.Background()); err != nil {
		t.Fatal(err)
	}
	if len(got) != 1 || len(got[0].Traces) != 1 {
		t.Fatalf("expected one batch with one trace, got %+v", got)
	}
	if keys[0] != "rgsa_test_abc" {
		t.Fatalf("missing key header: %q", keys[0])
	}
	tr := got[0].Traces[0]
	if tr["id"] == "" || tr["status"] != "success" || got[0].SDK["name"] != SDKName {
		t.Fatalf("unexpected trace %+v", tr)
	}
}

func TestSpanRecordsLatencyTemplateAndErrors(t *testing.T) {
	var got []captured
	var keys []string
	srv := server(t, 202, &got, &keys)
	defer srv.Close()
	c := New(Options{APIKey: "k", BaseURL: srv.URL, FlushInterval: time.Hour, DefaultMetadata: map[string]interface{}{"env": "test"}})
	msgs := []Message{{Role: "system", Content: "You are a ping bot."}, {Role: "user", Content: "ping"}}
	c.Start("gpt-4o-mini", OpenAI, msgs, WithTemplate("ping", "You are a ping bot."), WithMetadata(map[string]interface{}{"user": "u1"})).End("pong", 5, 1, nil)
	c.Start("gpt-4o", OpenAI, msgs).End("", 0, 0, errors.New("429 rate limited"))
	_ = c.Flush(context.Background())
	ok := got[0].Traces[0]
	if ok["output_text"] != "pong" || ok["prompt_tokens"].(float64) != 5 || ok["prompt_template"].(map[string]interface{})["name"] != "ping" {
		t.Fatalf("bad success trace %+v", ok)
	}
	if md := ok["metadata"].(map[string]interface{}); md["env"] != "test" || md["user"] != "u1" {
		t.Fatalf("metadata not merged: %+v", md)
	}
	bad := got[0].Traces[1]
	if bad["status"] != "error" || bad["error_message"] != "429 rate limited" || bad["prompt_template"].(map[string]interface{})["raw"] != "You are a ping bot." {
		t.Fatalf("bad error trace %+v", bad)
	}
	if _, has := bad["latency_ms"]; !has {
		t.Fatal("latency missing")
	}
}

func TestFlushAtTriggersSend(t *testing.T) {
	var got []captured
	var keys []string
	srv := server(t, 202, &got, &keys)
	defer srv.Close()
	c := New(Options{APIKey: "k", BaseURL: srv.URL, FlushAt: 2, FlushInterval: time.Hour})
	c.Trace(Trace{Model: "m", Provider: Other, InputMessages: []Message{{Role: "user", Content: "1"}}})
	c.Trace(Trace{Model: "m", Provider: Other, InputMessages: []Message{{Role: "user", Content: "2"}}})
	_ = c.Shutdown(context.Background())
	if len(got) != 1 || len(got[0].Traces) != 2 {
		t.Fatalf("expected auto flush of 2 traces, got %+v", got)
	}
}

func TestRetriesOnServerErrorAndReportsOnce(t *testing.T) {
	var hits int32
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { atomic.AddInt32(&hits, 1); w.WriteHeader(503) }))
	defer srv.Close()
	var reported int
	c := New(Options{APIKey: "k", BaseURL: srv.URL, FlushInterval: time.Hour, OnError: func(error) { reported++ }})
	c.Trace(Trace{Model: "m", Provider: Other, InputMessages: []Message{{Role: "user", Content: "x"}}})
	if err := c.Flush(context.Background()); err == nil {
		t.Fatal("expected error")
	}
	if atomic.LoadInt32(&hits) != 4 || reported != 1 {
		t.Fatalf("hits=%d reported=%d", hits, reported)
	}
}

func TestDisabledClientSendsNothing(t *testing.T) {
	var got []captured
	var keys []string
	srv := server(t, 202, &got, &keys)
	defer srv.Close()
	c := New(Options{APIKey: "k", BaseURL: srv.URL, Disabled: true})
	c.Trace(Trace{Model: "m", Provider: Other})
	_ = c.Shutdown(context.Background())
	if len(got) != 0 {
		t.Fatal("disabled client sent traces")
	}
}

File: packages/sdk-java/src/test/java/dev/regressa/RegressaTest.java
package dev.regressa;

import static org.junit.jupiter.api.Assertions.*;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class RegressaTest {
  private static final ObjectMapper JSON = new ObjectMapper();
  private HttpServer server;
  private final List<JsonNode> bodies = new CopyOnWriteArrayList<>();
  private final List<String> keys = new CopyOnWriteArrayList<>();
  private final AtomicInteger status = new AtomicInteger(202);
  private String url;

  @BeforeEach void up() throws Exception {
    server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext("/v1/traces", ex -> {
      bodies.add(JSON.readTree(ex.getRequestBody()));
      keys.add(ex.getRequestHeaders().getFirst(Regressa.KEY_HEADER));
      byte[] out = "{\"accepted\":1,\"rejected\":0,\"errors\":[]}".getBytes();
      ex.sendResponseHeaders(status.get(), out.length);
      ex.getResponseBody().write(out);
      ex.close();
    });
    server.start();
    url = "http://127.0.0.1:" + server.getAddress().getPort();
  }

  @AfterEach void down() { server.stop(0); }

  private Regressa client() { return Regressa.builder().apiKey("rgsa_test_abc").baseUrl(url).flushAt(10).flushInterval(Duration.ofHours(1)).build(); }

  @Test void bufferedTraceIsSentWithKeyHeader() {
    Regressa r = client();
    r.trace(Trace.builder("gpt-4o-mini", "openai").message("user", "hi").output("hello").build());
    r.flush();
    assertEquals(1, bodies.size());
    assertEquals("rgsa_test_abc", keys.get(0));
    JsonNode t = bodies.get(0).get("traces").get(0);
    assertEquals("hello", t.get("output_text").asText());
    assertEquals("success", t.get("status").asText());
    assertFalse(t.get("id").asText().isEmpty());
    assertEquals(Regressa.SDK_NAME, bodies.get(0).get("sdk").get("name").asText());
  }

  @Test void spanRecordsLatencyTemplateAndErrors() {
    Regressa r = Regressa.builder().apiKey("k").baseUrl(url).flushInterval(Duration.ofHours(1)).defaultMetadata(Map.of("env", "test")).build();
    r.start("gpt-4o-mini", "openai").message("system", "You are a ping bot.").message("user", "ping").template("ping", "You are a ping bot.").metadata(Map.of("user", "u1")).end("pong", 5, 1);
    r.start("gpt-4o", "openai").message("system", "Be brief.").message("user", "x").fail(new RuntimeException("429 rate limited"));
    r.flush();
    JsonNode ok = bodies.get(0).get("traces").get(0);
    assertEquals("pong", ok.get("output_text").asText());
    assertEquals(5, ok.get("prompt_tokens").asInt());
    assertEquals("ping", ok.get("prompt_template").get("name").asText());
    assertEquals("test", ok.get("metadata").get("env").asText());
    assertEquals("u1", ok.get("metadata").get("user").asText());
    assertTrue(ok.has("latency_ms"));
    JsonNode bad = bodies.get(0).get("traces").get(1);
    assertEquals("error", bad.get("status").asText());
    assertEquals("429 rate limited", bad.get("error_message").asText());
    assertEquals("Be brief.", bad.get("prompt_template").get("raw").asText(), "system prompt is inferred as the template");
  }

  @Test void flushAtTriggersSend() {
    Regressa r = Regressa.builder().apiKey("k").baseUrl(url).flushAt(2).flushInterval(Duration.ofHours(1)).build();
    r.trace(Trace.builder("m", "other").message("user", "1").build());
    r.trace(Trace.builder("m", "other").message("user", "2").build());
    r.shutdown();
    assertEquals(1, bodies.size());
    assertEquals(2, bodies.get(0).get("traces").size());
  }

  @Test void retriesServerErrorsAndReportsOnce() {
    status.set(503);
    AtomicInteger reported = new AtomicInteger();
    Regressa r = Regressa.builder().apiKey("k").baseUrl(url).flushInterval(Duration.ofHours(1)).onError(e -> reported.incrementAndGet()).build();
    r.trace(Trace.builder("m", "other").message("user", "x").build());
    r.flush();
    assertEquals(4, bodies.size(), "1 attempt + 3 retries");
    assertEquals(1, reported.get());
  }

  @Test void disabledClientSendsNothing() {
    Regressa r = Regressa.builder().apiKey("k").baseUrl(url).disabled(true).build();
    r.trace(Trace.builder("m", "other").build());
    r.shutdown();
    assertTrue(bodies.isEmpty());
  }
}
